src/shogun/machine/gp/KLDualInferenceMethod.cpp

```diff
@@ -507,14 +534,20 @@ void CKLDualInferenceMethod::update_alpha()
 
 float64_t CKLDualInferenceMethod::optimization()
 {
-	KLDualInferenceMethodMinimizer *minimizer=dynamic_cast<KLDualInferenceMethodMinimizer*>(m_minimizer);
-	REQUIRE(minimizer,"The minimizer must be an instance of KLDualInferenceMethodMinimizer\n"); 
+	CKLDualInferenceMethodMinimizer *minimizer=dynamic_cast<CKLDualInferenceMethodMinimizer*>(m_minimizer);
+	REQUIRE(minimizer,"The minimizer must be an instance of KLDualInferenceMethoMinimizer\n"); 
         KLDualInferenceMethodCostFunction* cost_fun=new  KLDualInferenceMethodCostFunction();
 	cost_fun->set_target(this);
+	bool cleanup=false;
+
+	if(this->ref_count()>1)
```

Comment:
Let me give you an example.
Case A: in KLDualInferenceMethod::method():
at time 1:  KLDualInferenceMethod: ref_count =0
at time 2:  cost_fun->set_target(this);   (KLDualInferenceMethod: ref_count =1)
at time 3:  SG_UNREF(cost_fun);  ((KLDualInferenceMethod: ref_count =0 and will delete KLDualInferenceMethod in a method in KLDualInferenceMethod  )



Case B: in KLDualInferenceMethod::method():
at time 1:  KLDualInferenceMethod: ref_count =1
at time 2:  cost_fun->set_target(this);   (KLDualInferenceMethod: ref_count =2)
at time 3:  SG_UNREF(cost_fun);  ((KLDualInferenceMethod: ref_count =1 and will NOT delete KLDualInferenceMethod in  )